docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,159 @@
+
+The motivation behind this proposal is to make all of this functionality accessible to clusters managed by Gardener.
+
+### Goals
+
+* Gardener must default to configuring its managed clusters with the CRI interface instead of the legacy Dockershim. This change must be transparent to Gardener users and must simply work out-of-the-box for new and existing clusters.
+* Low-level runtimes like gVisor or Kata Containers are provided as gardener extensions which are (optionally) installed into a landscape by the Gardener operator. There must be no runtime-specific knowledge in the core Gardener code.
+* It shall be possible to configure multiple low-level runtimes in Gardener clusters, on the Worker Group level.
+
+### Non-Goals
+
+* Exposing the configuration of the underlying runtimes to the user is beyond the scope of this proposal. The initial version will support a default common-sense configuration for each of the underlying runtimes. If required, future enhancements will make it possible to fine-tune the underlying runtimes.
+
+## Proposal
+
+Gardener today assumes that all supported operating systems have Docker pre-installed in the base image. Starting with Docker Engine 1.11, Docker itself was [refactored](https://www.docker.com/blog/docker-engine-1-11-runc/) and cleaned-up to be based on the [containerd](https://containerd.io/) library. The first phase would be to change the default Kubelet configuration as described [here](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd) so that Kubernetes would use containerd instead of the default Dockershim. This will be implemented for CoreOS, Ubuntu, and SuSE-JeOS. 
+
+Once Gardener clusters are migrated to use CRI, we will implement two Gardener extensions, providing gVisor and Kata Containers as options for Gardener landscapes. The `WorkerGroup` specification will be extended to allow specifying a list of additional required Runtimes for nodes in that group. For example:
+```yaml
+workers:
+- name: worker-b8jg5
+  machineType: m5.large
+  volumeType: gp2
+  volumeSize: 50Gi
+  autoScalerMin: 1
+  autoScalerMax: 2
+  maxSurge: 1
+  additionalRuntimes:
+  - gvisor
+  - kata-containers
+  machineImage:
+    name: coreos
+    version: 2135.6.0
+```
+
+Each extension will need to address the following concern:
+
+1. Add the low-level runtime binaries to the worker nodes. Each extension should store the runtime binary in a way that is accessible to a node to download and install during startup.
+1. Hook the runtime binary into the containerd configuration file, so that the runtime becomes available to containerd.
+1. Apply a label to each node that allows identifying nodes where the runtime is available.
```

Comment:
Can't this be done by Gardener directly when it configures the kubelets for the various worker pools, like here: https://github.com/gardener/gardener/blob/master/charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet.flags#L21-L25?

Reply:
Hi can you explain what do you mean? I didn't understand what exactly can be done similar to the init of the kubelet in the worker pools.

Reply:
I mean that this does not need to be done by the extension but can be done generically by Gardener itself as the worker pools will contain the `runtime` list.

Reply:
Concretely, we can  add the labels to the kubelet config here: https://github.com/gardener/gardener/blob/master/charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet.flags#L21-L25

Reply:
Correct i explained it in the design part 3.2.1